File: windows/Sources/ClearPowerCore/ProcessBudget.cs
// Estimate per-application power from CPU share x (package power - idle floor).
// Port of daemon/clearpowerd/sources/procs.py (attribution only; enumeration is platform code).
using System;
using System.Collections.Generic;
using System.Linq;

namespace ClearPower.Core
{
    public sealed class ProcessBudget
    {
        public double IntervalS { get; }
        private readonly double _floorWindowS;
        private double _next;
        private readonly LinkedList<(double t, double w)> _floor = new LinkedList<(double, double)>();
        public List<(string name, double w, double cpuPct)> Top { get; private set; } = new List<(string, double, double)>();

        public ProcessBudget(double intervalS = 2, double floorWindowS = 600)
        {
            IntervalS = intervalS;
            _floorWindowS = floorWindowS;
        }

        private double IdleFloor(double now, double packageW)
        {
            _floor.AddLast((now, packageW));
            while (_floor.Count > 0 && now - _floor.First!.Value.t > _floorWindowS) _floor.RemoveFirst();
            return _floor.Min(p => p.w);
        }

        /// <summary>`usage` yields (process name, cpu percent) pairs; evaluated at most every IntervalS.</summary>
        public List<(string name, double w, double cpuPct)> Sample(double now, double packageW, Func<IEnumerable<(string name, double cpuPct)>> usage, int n = 3)
        {
            if (now < _next) return Top;
            _next = now + IntervalS;
            var floor = packageW >= 0 ? IdleFloor(now, packageW) : 0.0;
            var budget = Math.Max(packageW - floor, 0.0);
            var agg = new Dictionary<string, double>();
            var order = new List<string>();
            var total = 0.0;
            foreach (var (name, c) in usage())
            {
                if (c <= 0) continue;
                total += c;
                var key = string.IsNullOrEmpty(name) ? "?" : name;
                if (!agg.ContainsKey(key)) { agg[key] = 0; order.Add(key); }
                agg[key] += c;
            }
            var top = new List<(string, double, double)>();
            if (total > 0)
            {
                // Counter.most_common: descending count, insertion order for ties.
                foreach (var key in order.OrderByDescending(k => agg[k]).Take(n))
                    top.Add((key, budget * agg[key] / total, agg[key]));
            }
            Top = top;
            return top;
        }
    }
}

File: macos/Sources/ClearPowerCore/ProcessBudget.swift
// Estimate per-application power from CPU share x (SoC power - idle floor).
// Port of daemon/clearpowerd/sources/procs.py minus psutil.
import Foundation

public struct ProcessBudget {
    public let interval: Double
    public let floorWindow: Double
    private var next = 0.0
    private var floor: [(t: Double, w: Double)] = []
    public private(set) var top: [(name: String, w: Double, cpuPct: Double)] = []

    public init(intervalS: Double = 3, floorWindowS: Double = 600) {
        interval = intervalS
        floorWindow = floorWindowS
    }

    private mutating func idleFloor(now: Double, packageW: Double) -> Double {
        floor.append((now, packageW))
        while let f = floor.first, now - f.t > floorWindow { floor.removeFirst() }
        return floor.map { $0.w }.min() ?? 0
    }

    public var due: Bool { true }

    /// `usage` is (process name, cpu percent) for every process with cpu > 0. Returns the
    /// top-n names with their share of the budget. Returns the previous result if called
    /// before `interval` has elapsed (pass `force` to bypass).
    public mutating func sample(now: Double, packageW: Double, usage: () -> [(String, Double)],
                                n: Int = 3, force: Bool = false) -> [(name: String, w: Double, cpuPct: Double)] {
        if now < next && !force { return top }
        next = now + interval
        let fl = packageW >= 0 ? idleFloor(now: now, packageW: packageW) : 0
        let budget = max(packageW - fl, 0)
        var agg: [String: Double] = [:]
        var total = 0.0
        for (name, c) in usage() where c > 0 {
            total += c
            agg[name.isEmpty ? "?" : name, default: 0] += c
        }
        var out: [(name: String, w: Double, cpuPct: Double)] = []
        if total > 0 {
            for (name, c) in agg.sorted(by: { $0.value > $1.value }).prefix(n) {
                out.append((name, budget * c / total, c))
            }
        }
        top = out
        return out
    }
}
